aten/src/ATen/native/Correlation.cpp

```diff
@@ -0,0 +1,107 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+namespace at {
+namespace native {
+
+Tensor cov(
+    const Tensor& self,
+    int64_t correction,
+    const c10::optional<Tensor>& fweights,
+    const c10::optional<Tensor>& aweights) {
+  constexpr int OBSERVATIONS_DIM = 1;
+
+  TORCH_CHECK(
+      self.ndimension() <= 2,
+      "cov(): expected input to have two or fewer dimensions but got an input with ",
+      self.ndimension(),
+      " dimensions");
+
+  TORCH_CHECK(
+      self.scalar_type() != kBool, "cov(): bool dtype is not supported for input");
+
+  // View input tensor as 2D (variables, observations)
+  auto in = self.ndimension() < 2 ? self.view({1, -1}) : self;
+  const auto num_observations = in.size(OBSERVATIONS_DIM);
+
+  // The product of frequencies (fweights) and weights (aweights).
+  Tensor w;
+
+  if (fweights.has_value()) {
+    w = fweights.value();
+    TORCH_CHECK(
+        w.ndimension() <= 1,
+        "cov(): expected fweights to have one or fewer dimensions but got fweights with ",
+        w.ndimension(),
+        " dimensions");
+    TORCH_CHECK(
+        at::isIntegralType(w.scalar_type(), false),
+        "cov(): expected fweights to have integral dtype but got fweights with ",
+        w.scalar_type(),
+        " dtype");
+    TORCH_CHECK(
+        w.numel() == num_observations,
+        "cov(): expected fweights to have the same numel as there are observations in the input but got ",
+        w.numel(),
+        " != ",
+        num_observations);
+    TORCH_CHECK(
+        !(w.is_cpu() && w.lt(0).any().item<bool>()),
```

Comment:
@ngimel Should I be skipping this test on CUDA to avoid an implicit cross-device sync? What is the current thinking on this?

Reply:
No. Cross-device syncs are ok (in fact, all tests are doing them, repeatedly), we are skipping (typically) tests for the cases where invalid inputs cause device side asserts.